### ISSUE TITLE
Homekit automations don't seem to work reliably.

### ISSUE THREAD

**Issue by rcjcarr** [grivkees/homebridge-carrier-infinity#147]
**Describe The Bug:**
This is about Apple Homekit  *Automations* not running reliably with this plugin. With *manual* Homekit  'Thermostat' accessory adjustments, the plugin works fine. Both the Homekit and he Carrier Infinity wall thermostats respond as expected (new setpoint), and no error logs are generated in HOOBS logs. I also have no problems setting the Homekit Thermostat accessory with Siri voice commands (using Apple HomePod mini), and this does not generate HOOBS/Homebridge log entries either.

The problematic automations consist of setting the Homekit Thermostat accessory temperature at a given time of day. The first few automations seem to work, sometimes even with no error messages added to the HOOBS log.  Both the Homekit Thermostat accessory and the Carrier Infinity wall thermostat register the new temperature setpoint. After of few of these, Apple Homekit temperature automations have no effect on either thermostats, with HOOBS/HomeBridge log ERROR messages shown below appearing at the given automation trigger times.

Also I tried adding a few extra parameters to the JSON config info (temperature units, zone, bridge details) as shown below but this had no effect.

**To Reproduce:**
Create an Apple Homekit automation --on the thermostat accessory-- to set the thermostat to a certain temperature at a certain time.

**Expected behavior:**
Thermostat temperature setpoint should update.

**Logs:**
*(Some data redacted with xxx)*

At HOOBS Bridge startup:
```
1/14/2022. 10:42:16 CarrierInfinity Discovered device serial:25x1xxxxx794 zone:0
1/14/2022, 10:42:16 AMHoobsCarrierInfinityDiscovered environmental sensor device serial:251xxxxxx794
1/14/2022, 10:42:17 AMHoobsCarrierInfinityAdding new accessory:  ZONE 1 Environmental Sensor
1/14/2022, 10:42:17 AMHoobsCarrierInfinityDiscovered outdoor temp sensor device serial:25xxxxxx94
1/14/2022, 10:42:17 AMHoobsCarrierInfinityAdding new accessory:  Outdoor Temperature
```

**When bug presents itself (automation not working):**

```
ERROR[Hoobs F6AB@Cooling Threshold Temperature] The write handler for the characteristic 'Cooling Threshold Temperature' on the accessory 'ZONE 1' was slow to respond!
ERROR[Hoobs F6AB@Heating Threshold Temperature] The write handler for the characteristic 'Heating Threshold Temperature' on the accessory 'ZONE 1' was slow to respond!
```

**Sometimes this is shown instead:**
```
ERROR[Hoobs F6AB@Target Temperature] The write handler for the characteristic 'Target Temperature' on the accessory 'ZONE 1' was slow to respond!
ERROR[Hoobs F6AB@Target Temperature] The write handler for the characteristic 'Target Temperature' on the accessory 'ZONE 1' was slow to respond!
ERROR[Hoobs F6AB@Heating Threshold Temperature] The write handler for the characteristic 'Heating Threshold Temperature' on the accessory 'ZONE 1' was slow to respond!
ERROR[Hoobs F6AB@Heating Threshold Temperature] The write handler for the characteristic 'Heating Threshold Temperature' on the accessory 'ZONE 1' was slow to respond!
```

**Plugin Config:**
```
{
    "platform": "CarrierInfinity",
    "username": "xxxxxxxxxx",
    "password": "xxxxxxxxxx",
    "showOutdoorTemperatureSensor": false,
    "showFanControl": true,
    "showIndoorHumiditySensors": false,
    "holdBehavior": "forever",
    "tempUnits": "c",
    "zones": [
        {
            "id": "1",
            "fan": "off"
        }
    ]
}
```

**Screenshots:**

**Environment:**

* **Plugin Version**: 1.5.4
* **Homebridge Version**: 1.3.9
* **Node.js Version**: 16.13.2
* **NPM Version**: 8.1.2
* **Operating System**: 
Raspbian (HOOBs box, s/w version 4.1.21):
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)" NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"

**Comment by rcjcarr** [comment]
**Update**
I changed the Plugin Configuration to:
```
{
    "platform": "CarrierInfinity",
    "username": "xxxxxxxxxx",
    "password": "xxxxxxxxx",
    "showOutdoorTemperatureSensor": true,
    "showFanControl": false,
    "showIndoorHumiditySensors": false,
    "holdBehavior": "forever"
}
```

Having the 'showFanControl' set to 'false' seems to resolve it... will wait a few days to see.

**Comment by rcjcarr** [comment]
## This morning's results
Thermostat did NOTchange to new setpoint:
```
1/16/2022, 5:00:02 AMCarrier Infinity BridgeERROR[Carrier Infinity Bridge E27F@Target Temperature] The write handler for the characteristic 'Target Temperature' on the accessory 'ZONE 1' was slow to respond!
1/16/2022, 5:00:02 AMCarrier Infinity BridgeERROR[Carrier Infinity Bridge E27F@Cooling Threshold Temperature] The write handler for the characteristic 'Cooling Threshold Temperature' on the accessory 'ZONE 1' was slow to respond!
```
Thermostat did NOT change to new setpoint:
```
1/16/2022, 6:00:02 AMCarrier Infinity BridgeERROR[Carrier Infinity Bridge E27F@Cooling Threshold Temperature] The write handler for the characteristic 'Cooling Threshold Temperature' on the accessory 'ZONE 1' was slow to respond!
1/16/2022, 6:00:02 AMCarrier Infinity BridgeERROR[Carrier Infinity Bridge E27F@Target Temperature] The write handler for the characteristic 'Target Temperature' on the accessory 'ZONE 1' was slow to respond!
```

Thermostat DID change to new setpoint!:
```
1/16/2022, 7:00:02 AMCarrier Infinity BridgeERROR[Carrier Infinity Bridge E27F@Cooling Threshold Temperature] The write handler for the characteristic 'Cooling Threshold Temperature' on the accessory 'ZONE 1' was slow to respond!
1/16/2022, 7:00:02 AMCarrier Infinity BridgeERROR[Carrier Infinity Bridge E27F@Heating Threshold Temperature] The write handler for the characteristic 'Heating Threshold Temperature' on the accessory 'ZONE 1' was slow to respond!
```

Definitely looks like an error with the write handler for the characteristic 'Target Temperature.' Not sure where this is happening, plugin level, Homebridge level, or Homekit.

**Comment by rcjcarr** [comment]
Changed JSON config file to:
```
{
    "accessories": [],
    "platforms": [
        {
            "platform": "CarrierInfinity",
            "username": "xxxxxxxxx",
            "password": "xxxxxxxxx",
            "showOutdoorTemperatureSensor": true,
            "showFanControl": false,
            "showIndoorHumiditySensors": false,
            "holdBehavior": "forever"
        }
    ],
    "bridge": {
        "name": "Carrier Infinity Bridge",
        "username": "4D:6C:20:5C:6F:B5",
        "port": 51826,
        "pin": "031-45-154"
    }
}
```
with no improvement.

**Comment by rcjcarr** [comment]
I cycled the powermy HomePod Mini;

**Comment by rcjcarr** [comment]
The good folks at HOOBS indicated:

> We checked the plugin, and it seems there's an issue within the answer time of the device.
This could be resolved by the plugin developer in raising the answertime to avoid the "'ZONE 1' was slow to respond!"

It looks like timeout occurs after about 2s. Maybe raise this by a factor of 5 or 10?

**Comment by rcjcarr** [comment]
A workaround consists of creating 'Scenes' for every desired temperature setting and mode (heat/cool), and scheduling these 'Scenes' in the Apple Home app. This does not generate any HOOBS log entries (so far), and seems to work 'the first time, every time, all the time.' New temperature setpoints also appear fairly quickly in the Carrier MyInfinity Toucn user portal (https://www.myinfinitytouch.carrier.com) . 

I'm fairly new to Apple Homekit and it would appear this is the preferred approach, *i.e.*, create 'Scenes' and schedule them, as opposed to assigning automations directly to accessories.

In conclusion it feels like the Apple Home App (and how it deals with *accessory* automations) is causing something to go awry when interacting with this plugin. Fault attribution could conceivably by entirely with the Home App and not the plugin.

**Comment by rcjcarr** [comment]
Update:
Even with defined 'scenes' as mentioned above I'm still getting Fan Control and Temperature setpoint issues with the Carrier plugin with HOOBS. This issue does not seem occur running Homebridge (OS= Ubuntu right now). It only appears when I use HOOBS. Either the HOOBS Strawberry Pi h/w issue (processor not fast enough and/or otherwise lacking something)  or HOOBS s/w issue.

**Comment by rcjcarr** [comment]
No issues were observed with this plugin running Homebridge on the HOOBS Box (Raspberry Pi 3B+) running Homebridge directly (changed the SD card).

**Comment by grivkees** [comment]
Hey! Thanks for writing this up!
Can you confirm if you have a schedule set or not on the thermostat itself?

I actually have never tried using HomeKit automations. I’ll give it a try and see if I can figure out what’s going on.

**Comment by rcjcarr** [comment]
Hi Glenn,
I was still getting 'ignored'  Homekit (HK) automations using straight Homebridge, both running on a Pi 3B+ and a Dell Optiplex Desktop running Ubuntu 20. I venture to guess that HOOBS is more explicit in presenting error messages in its logs compared to straight Homebridge so I have reverted back to HOOBS as my HK bridge of choice. 

To answer your question, the thermostat is on permanent 'Hold' at all times when running your pluging, i.e. JSON holdBehavior="forever". I'm still getting the odd error as follows, mostly related to fan speed/mode changes, but also sometimes as a result of temperature change automations:
```
2/14/2022, 9:00:03 PMCarrier Infinity BridgeERROR[Carrier Infinity Bridge BC0E@Rotation Speed] The write handler for the characteristic 'Rotation Speed' on the accessory 'ZONE 1' was slow to respond!
2/15/2022, 4:45:03 AMCarrier Infinity BridgeERROR[Carrier Infinity Bridge BC0E@Target Fan State] The write handler for the characteristic 'Target Fan State' on the accessory 'ZONE 1' was slow to respond!
```
Commanding thermostat changes using HK 'scenes' (as opposed to direct accessory control via Apple Home app, this latter feature unique to the Apple Home app) seems more reliable but not 100%. Oddly enough commanding thermostat changes using Siri voice commands seems more reliable still. For a while I was questioning HK hub (Apple TV 4k vs Homepod mini) and I cannot attribute blame to any of them at this point. Creating HK automations via the 'Home+5' app seems to be more reliable than creating automations in Apple 'Home' or the 'Controller for Homekit' app. Not quite sure what's going on there.

It looks like the  comment from the folks at HOOBS above may be the closest to what's going on, *i.e.*, an issue with the answer time of the 'device,' which I'm guessing would the the Carrier server/API. I'm speculating that the plugin needs to wait longer, or confirm with the API that it has accepted the setting, and/or have the plugin  retry if not? Maybe the plugin does this already?

Anyways sorry about my ramblings/ruminations. I appreciate any time you would put into this. The Carrier Infinity thermostat is unique in that it communicates with DC pulses to the furnace (I2C perhaps), as opposed to straight relay contacts. So I can't replace it with an Ecobee thermostat for example hence my appreciation of your plugin. The problem with the Carrier 'app' (and thermostat) is that it only accepts 4 schedules, and no HK support, hence my use of your plugin.

**Comment by grivkees** [comment]
Thanks for all the context!

Re: the "was slow to respond!" errors. That is tracked in #111. It's a real issue, but I'm skeptical thats the root issue here though.

I have seen some instances myself where I can set a temperature via HomeKit, but it doesn't 'stick'. I think theres an edge case where the carrier api accepts the update, but it never makes it to the thermostat since the thermostat thinks it has a more recent change. My guess is thats the issue here. I'll dig into it some more.

> The Carrier Infinity thermostat is unique in that it communicates with DC pulses to the furnace (I2C perhaps), as opposed to straight relay contacts. So I can't replace it with an Ecobee thermostat for example hence my appreciation of your plugin. The problem with the Carrier 'app' (and thermostat) is that it only accepts 4 schedules, and no HK support, hence my use of your plugin.

Yep, this is why I made the plugin in the first place. Can't replace the thermostat without losing features. But I really wanted drop in HK support. (If you're curious, there's actually some details on how they communicate on the [infinitive](https://github.com/acd/infinitive#details) project.) I hadn't actually thought about using HK to get around the schedule limitations, but its a really good idea, and something that should definitely work.

**Comment by rcjcarr** [comment]
Very interesting. Yesterday I emailed a local Carrier dealer (in Canada):
> I noticed that Carrier has joined forces with Ecobee to some extent:
 https://www.carrier.com/residential/en/ca/products/thermostats/smart-thermostats/ecobee-smart-thermostats-powered-by-carrier/  
Is this thermostat compatible with my modulating furnace? Is it a ‘communicating’ type, similar to the Infinity Touch? Standard Ecobee thermostats only seem to offer 2 fan speeds and 2 burner settings for example, which would not be a good match for my furnace…

He responded:
>It is not compatible, infinity controls communicate directly to infinity equipment via DC pulse signals , all other thermostats are wifi switches .

**Comment by rcjcarr** [comment]
A few more observations after of few more weeks of monitoring:
* Automated Homekit temperature _changes_ (via HK scenes) seems to work now. These generate the odd error message in the HOOBS log (see below)  but these are seemingly inconsequential as far as the end result is concerned (temperature setting are correct/as expected). 
* Automated Homekit fan settings _changes_ (via HK scenes) from MANUAL (any speed) to 'AUTO' also work. 
* Automated Homekit fan setting _changes_ (via HK scenes) from 'AUTO' to  'MANUAL/LOW' don't work; the fan mode goes to MANUAL but the 'MEDIUM' fan speed is was is achieved instead of the expected 'LOW.'
```
2/28/2022, 9:00:04 PMCarrier Infinity BridgeERROR[Carrier Infinity Bridge BC0E@Target Fan State] The write handler for the characteristic 'Target Fan State' on the accessory 'ZONE 1' was slow to respond!
2/28/2022, 9:00:04 PMCarrier Infinity BridgeERROR[Carrier Infinity Bridge BC0E@Rotation Speed] The write handler for the characteristic 'Rotation Speed' on the accessory 'ZONE 1' was slow to respond!
3/1/2022, 7:44:59 AMWARNINGplugin site unavailable
3/1/2022, 10:00:02 AMCarrier Infinity BridgeERROR[Carrier Infinity Bridge BC0E@Target Heating Cooling State] The read handler for the characteristic 'Target Heating Cooling State' on the accessory 'ZONE 1' was slow to respond!
3/1/2022, 10:00:02 AMCarrier Infinity BridgeERROR[Carrier Infinity Bridge BC0E@Target Temperature] The read handler for the characteristic 'Target Temperature' on the accessory 'ZONE 1' was slow to respond!
```

**Comment by grivkees** [comment]
Just pushed some changes to the "next" channel that should hopefully make updates more reliable and handle some of the race conditions you were seeing.

Give it a try when you get a chance and let me know if things seem better, or if you see any bugs. (Especially if you see the device in homekit stop responding or hang.)

npm install -g homebridge-carrier-infinity@next

**Comment by rcjcarr** [comment]
OK just installed v1.5.7 on my HOOBS box (Rpi 3B+). Will monitor for a while and give you feedback. Thanks again for looking into this, that's fantastic.

**Comment by rcjcarr** [comment]
Just saw this after a manual temperature change at 10AM:
```
3/3/2022, 9:51:39 AMCarrier Infinity BridgeLoaded plugin 'homebridge-carrier-infinity'
3/3/2022, 9:51:39 AMCarrier Infinity BridgeLoading 1 platforms...
3/3/2022, 9:51:40 AMCarrier Infinity BridgeBridge is running on port 51826.
3/3/2022, 9:51:43 AMCarrier Infinity BridgeCarrierInfinityDiscovered device serial:2514W003794 zone:0
3/3/2022, 9:51:44 AMCarrier Infinity BridgeCarrierInfinityDiscovered environmental sensor device serial:2514W003794
3/3/2022, 9:51:44 AMCarrier Infinity BridgeCarrierInfinityDiscovered outdoor temp sensor device serial:2514W003794
3/3/2022, 9:51:39 AMDummy BridgeLoaded plugin 'homebridge-dummy'
3/3/2022, 9:51:39 AMDummy BridgeLoading 2 accessories...
3/3/2022, 9:51:40 AMDummy BridgeTEMP HOLDSetting switch to true
3/3/2022, 9:51:41 AMDummy BridgeBridge is running on port 51836.
3/3/2022, 10:00:42 AMCarrier Infinity BridgeERROR[Carrier Infinity Bridge BC0E@Target Temperature] The write handler for the characteristic 'Target Temperature' on the accessory 'ZONE 1' was slow to respond!
```

**Comment by rcjcarr** [comment]
The above error could have been due to two temperature changes made in too close succession (I was playing with the HK thermostat temperature setting and was having finger problems).

Also worth a mention: Last night for the first time the FAN setting went to MANUAL LOW as expected, instead of MANUAL-MED. I reflashed my HOOBS micro SD card earlier in the day for other reasons, not sure if this was a factor.

**Comment by rcjcarr** [comment]
```
3/3/2022, 9:51:03 PMCarrier Infinity BridgeERROR[Carrier Infinity Bridge BC0E@Rotation Speed] The write handler for the characteristic 'Rotation Speed' on the accessory 'ZONE 1' was slow to respond!
3/3/2022, 9:54:03 PMCarrier Infinity BridgeERROR[Carrier Infinity Bridge BC0E@Target Fan State] The write handler for the characteristic 'Target Fan State' on the accessory 'ZONE 1' was slow to respond!
3/4/2022, 4:45:03 AMCarrier Infinity BridgeERROR[Carrier Infinity Bridge BC0E@Target Fan State] The write handler for the characteristic 'Target Fan State' on the accessory 'ZONE 1' was slow to respond!
3/4/2022, 5:00:03 AMCarrier Infinity BridgeERROR[Carrier Infinity Bridge BC0E@Target Temperature] The write handler for the characteristic 'Target Temperature' on the accessory 'ZONE 1' was slow to respond!
```

**Comment by rcjcarr** [comment]
OK got back to this today. The 'next' channel version (1.5.7) I could **not** install on HOOBS (on a Pi 3B+). I was able to install it on Homebridge (both the aforementioned Pi and on a separate Ubuntu machine, on at a time of course). For both of these, v1.5.7 works at first (temperature settings pushed on Carrier API successfully),  but as soon as a fan setting is commanded, it locks up (dreaded 'No Response')  on HK tiles from either the thermostat setting or the fan setting. Only recovery at that point is to re-start Homebridge.

**Comment by rcjcarr** [comment]
Both HK Thermostat and Fan tiles also show 'No Response' if temperature changes are commanded too quickly.

**Comment by rcjcarr** [comment]
| Initial Fan State | Final State AUTO (manually requested by Home app)  | Final State LOW (manually requested by Home app)  | Final State MED (manually requested by Home app) | Final State HIGH (manually requested by Home app) |
| :---: | :-----------: | :-------: | :--------: | :-------: |
| AUTO | - |  hangs  | tbd|  tbd |
| LOW | OK |  - | tbd |  tbd |
| MED | OK |  tbd | - |  tbd |
| HIGH | OK |  tbd | tbd |  - |

**Comment by rcjcarr** [comment]
Getting this once in a while, and losing control from my iPhone. Only recourse is to restart the plugin.
```
[11/03/2022, 13:54:31] [homebridge-carrier-infinity] Could not refresh api access token:  getaddrinfo ENOTFOUND www.app-api.ing.carrier.com 
Status:  undefined 
Data:  undefined
[11/03/2022, 13:54:31] [homebridge-carrier-infinity] Failed to fetch updates (axios):  getaddrinfo ENOTFOUND www.app-api.ing.carrier.com
[11/03/2022, 13:54:31] [homebridge-carrier-infinity] Could not discover devices: Cannot read properties of null (reading 'locations')
```

**Comment by rcjcarr** [comment]
```
[14/03/2022, 17:32:30] [homebridge-carrier-infinity] This plugin slows down Homebridge. The read handler for the characteristic 'Target Temperature' was slow to respond! See https://git.io/JtMGR for more info.
[14/03/2022, 17:32:30] [homebridge-carrier-infinity] This plugin slows down Homebridge. The read handler for the characteristic 'Cooling Threshold Temperature' was slow to respond! See https://git.io/JtMGR for more info.
[14/03/2022, 17:32:30] [homebridge-carrier-infinity] This plugin slows down Homebridge. The read handler for the characteristic 'Heating Threshold Temperature' was slow to respond! See https://git.io/JtMGR for more info.
[14/03/2022, 17:32:36] [homebridge-carrier-infinity] This plugin slows down Homebridge. The read handler for the characteristic 'Target Temperature' didn't respond at all!. Please check that you properly call the callback! See https://git.io/JtMGR for more info.
[14/03/2022, 17:32:36] [homebridge-carrier-infinity] This plugin slows down Homebridge. The read handler for the characteristic 'Cooling Threshold Temperature' didn't respond at all!. Please check that you properly call the callback! See https://git.io/JtMGR for more info.
[14/03/2022, 17:32:36] [homebridge-carrier-infinity] This plugin slows down Homebridge. The read handler for the characteristic 'Heating Threshold Temperature' didn't respond at all!. Please check that you properly call the callback! See https://git.io/JtMGR for more info.
```
Further testing suggests that the above errors (since v1.5.7) happen under 2 conditions:
1) The Carrier Infinity plugin is running as a child bridge
2) and a network change occurs (WiFi-->LTE-->Wifi)

Running the Carrier Infinity plugin under Homebridge directly (*i.e.*, **not** as a child bridge)  does not seem to produce these errors and seems much more reliable.

**Comment by rcjcarr** [comment]
I think this got fixed with the iOS 15.4 update today.(!)

**Comment by rcjcarr** [comment]
Another example: this after a Wifi-LTE-Wifi change (going from work to home) and adjusting the temperature once at home.
```
[15/03/2022, 15:47:08] [CarrierInfinity] Changes sent to carrier api successfully.
[15/03/2022, 17:27:55] Error: send ENETUNREACH 224.0.0.251:5353
    at doSend (node:dgram:722:16)
    at defaultTriggerAsyncIdScope (node:internal/async_hooks:465:18)
    at afterDns (node:dgram:668:5)
    at processTicksAndRejections (node:internal/process/task_queues:84:21)
[15/03/2022, 17:27:55] Error: send ENETUNREACH 224.0.0.251:5353
    at doSend (node:dgram:722:16)
    at defaultTriggerAsyncIdScope (node:internal/async_hooks:465:18)
    at afterDns (node:dgram:668:5)
    at processTicksAndRejections (node:internal/process/task_queues:84:21)
[15/03/2022, 17:27:55] Error: send ENETUNREACH 224.0.0.251:5353
    at doSend (node:dgram:722:16)
    at defaultTriggerAsyncIdScope (node:internal/async_hooks:465:18)
    at afterDns (node:dgram:668:5)
    at processTicksAndRejections (node:internal/process/task_queues:84:21)
[15/03/2022, 17:27:55] Error: send ENETUNREACH 224.0.0.251:5353
    at doSend (node:dgram:722:16)
    at defaultTriggerAsyncIdScope (node:internal/async_hooks:465:18)
    at afterDns (node:dgram:668:5)
    at processTicksAndRejections (node:internal/process/task_queues:84:21)

**Comment by grivkees** [comment]
@rcjcarr thanks for testing this!
I'm fairly confident I found and fixed the deadlock in the new logic that was causing the hang.
Please give the new 'next' version a shot when you get a chance, and reopen the issue if you're still seeing issues.
